### ISSUE TITLE
semver_index.py algorithm doesn't seem to support non-standard versions

### ISSUE THREAD

**Issue by jimshowalter** [google/osv.dev#1953]
**Describe the bug**

Having converted semver_index.py to Java, including the unit tests, everything worked fine until we tried to use it to match versions in ranges, where the versions are from the .json vuln files instead of from tests.

The .json vuln files turn out to be messier than the tests.

For example, bouncy-castle's bc-fips jar has version 1.0.2.1, and GHSA-rpr3-cw39-3pxh.json lists 2.9.10.1, 2.9.10.2, 2.9.10.3, 2.9.9.1, 2.9.9.2, 2.9.9.3, etc.

Adding 1.0.2.1 to semver_index_test.test_normalize results in a failure:

```
    def test_normalize(self):
        """Test version normalization."""
        versions = [
            '1.0.2.1', <<< ADDED
            '1.0.0-beta.2+BLAH', '1.0.0-beta.2', '1.0.0-beta.11', '1.0.0-rc.1',
            '1.0.0', '1.0.0-alpha', '1.0.0-alpha.1', '1.0.0-alpha.beta',
            '1.0.0-beta', '1.0.0-9', '1.0.0-0a',
            '24.4.3001-20191109021931-daa7c04131f5'
        ]

        self.assertListEqual([
            '', <<< ADDED
            '00000001.00000000.00000000-1beta.00000002',
            '00000001.00000000.00000000-1beta.00000002',
            '00000001.00000000.00000000-1beta.00000011',
            '00000001.00000000.00000000-1rc.00000001',
            '00000001.00000000.00000000-zzzzzzzzzzzzzzzz',
            '00000001.00000000.00000000-1alpha',
            '00000001.00000000.00000000-1alpha.00000001',
            '00000001.00000000.00000000-1alpha.1beta',
            '00000001.00000000.00000000-1beta',
            '00000001.00000000.00000000-00000009',
            '00000001.00000000.00000000-10a',
            '00000024.00000004.00003001-120191109021931-daa7c04131f5'
        ], [semver_index.normalize(version) for version in versions])
```

```
Ran 1 test in 0.003s

FAILED (errors=1)

Error
Traceback (most recent call last):
  File "/Users/jshowalter/code/other/osv.dev/gcp/api/osv/semver_index_test.py", line 107, in test_normalize
    ], [semver_index.normalize(version) for version in versions])
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jshowalter/code/other/osv.dev/gcp/api/osv/semver_index_test.py", line 107, in <listcomp>
    ], [semver_index.normalize(version) for version in versions])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jshowalter/code/other/osv.dev/gcp/api/osv/semver_index.py", line 110, in normalize
    version = parse(version)
              ^^^^^^^^^^^^^^
  File "/Users/jshowalter/code/other/osv.dev/gcp/api/osv/semver_index.py", line 104, in parse
    return semver.Version.parse(coerce(version))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jshowalter/code/other/osv.dev/.venv/lib/python3.11/site-packages/semver/version.py", line 646, in parse
    raise ValueError(f"{version} is not valid SemVer string")
ValueError: 1.0.2.1 is not valid SemVer string
```

So how can this be working?: https://osv.dev/list?ecosystem=&q=org.bouncycastle%3Abc-fips

https://osv.dev/vulnerability/GHSA-68m8-v89j-7j2p

If a project happens to be using 1.0.2.1, it won't be detected because the range matching doesn't work (it would blow up with the same error about invalid SemVer).

**Comment by another-rex** [comment]
> So how can this be working?: https://osv.dev/list?ecosystem=&q=org.bouncycastle%3Abc-fips

This works because most Java packages are not categorized as a SemVer ecosystem, you can see if it's semver by looking at this field: https://ossf.github.io/osv-schema/#affectedrangestype-field. Specifically this advisory is classified as `ECOSYSTEM` versioning.

```
      "ranges": [
        {
          "type": "ECOSYSTEM", // <---
          "events": [
            {
              "introduced": "0"
            },
            {
              "fixed": "1.0.2.4"
            }
          ]
        }
      ]
```

For every non-semver ecosystem, we implement custom sorting here: https://github.com/google/osv.dev/tree/master/osv/ecosystems

We also do not support generic range matching for non-semver ecosystems, but we do enumerate through all published versions of the package and index them to do exact matching.

**Comment by jimshowalter** [comment]
Ah hah!  Thank you for the quick reply.
